src/forwardproblem.cpp

```diff
@@ -109,25 +130,65 @@ void ForwardProblem::workForwardProblem() {
 
     /* store initial state and sensitivity*/
     initial_state_ = getSimulationState();
+    // store root information at t0
+    model->froot(t_, x_, dx_, rootvals_);
+
+    // get list of trigger timepoints for fixed-time triggered events
+    auto trigger_timepoints = model->get_trigger_timepoints();
+    auto it_trigger_timepoints = std::find_if(
+        trigger_timepoints.begin(), trigger_timepoints.end(),
+        [this](auto t) { return t > this->t_; }
+    );
 
     /* loop over timepoints */
     for (it_ = 0; it_ < model->nt(); it_++) {
-        auto nextTimepoint = model->getTimepoint(it_);
+        // next output time-point
+        auto next_t_out = model->getTimepoint(it_);
 
-        if (std::isinf(nextTimepoint))
+        if (std::isinf(next_t_out))
             break;
 
-        if (nextTimepoint > model->t0()) {
-            // Solve for nextTimepoint
-            while (t_ < nextTimepoint) {
-                int status = solver->run(nextTimepoint);
-                solver->writeSolution(&t_, x_, dx_, sx_, dx_);
+        if (next_t_out > model->t0()) {
+            // Solve for next output timepoint
+            while (t_ < next_t_out) {
+                if (is_next_t_too_close(t_, next_t_out)) {
+                    // next timepoint is too close to current timepoint
+                    // update `t_`, required by `handleDataPoint`
+                    t_ = next_t_out;
```

Comment:
isnt the value of `t_` always overwritten by `solver->writeSolution`?

Reply:
Correct. But `solver->writeSolution` is only called for the next timepoint once `handleDataPoint` finishes. By then it's fine that it gets overwritten.

Reply:
That‘s not evident from the code here though